pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -385,10 +416,45 @@ func (s *eventStream) addSST(
 	// TODO(casper): add metrics to track number of SSTs, and number of ssts
 	// that are not inside the boundaries (and possible count+size of kvs in such ssts).
 	size := 0
-	if err := s.trimSST(sst, func(keyVal roachpb.KeyValue) {
-		batch.KeyValues = append(batch.KeyValues, keyVal)
-		size += keyVal.Size()
-	}); err != nil {
+	timestampToDelRanges := make(map[hlc.Timestamp][]*roachpb.RangeFeedDeleteRange)
+	// Iterator may release fragmented ranges, we try to de-fragment them
+	// before we release roachpb.RangeFeedDeleteRange events.
+	mergeDelRange := func(unsafeRangeKey storage.MVCCRangeKey) {
+		delRanges, ok := timestampToDelRanges[unsafeRangeKey.Timestamp]
+		if ok {
+			for _, delRange := range delRanges {
```

Comment:
Since we iterate across and add range keys in order (left-to-right in the key space), I believe we can just check the last member's end key here rather than iterating over all of them, which avoids making this quadratic.